### DISCUSSION TITLE
Do API tokens expire?

### DISCUSSION THREAD

**Discussion by cocowalla** [umami-software/umami#1170] [Q&A]
The docs explain how to authenticate and obtain a bearer token for use with the API - but do the obtained tokens expire?

If so, after how long? Does the API return a particular response for expired tokens?

Thanks!

**Comment by bekanne** [comment]
@cocowalla did you find out something?

**Comment by cocowalla** [comment]
@bekanne No, and the tokens returned by Umami aren't regular JWTs, so I've no idea how to even decode them. [Docs](https://umami.is/docs/api/authentication) still don't say anything about expiry time either 🤷 

EDIT: I also had a quick look at the code. I'm not really a TypeScript kinda guy, and I know nothing of Next.js, so it's quite possible I'm missing something, but AFAIKT, here is what I found.

When logging in, it hits [saveAuth](https://github.com/umami-software/umami/blob/master/src/lib/auth.ts#L14), which points to [createSecureToken](https://github.com/umami-software/next-basics/blob/fc0dcd54cd3d08196b6229053d030ae82486bd66/src/server/token.ts#L16), which in turn uses [createToken](https://github.com/umami-software/next-basics/blob/master/src/server/token.ts#L4) to create a JWT... and then... strangely appears to encrypt the entire JWT using AES-256-GCM, resulting in a totally opaque blob?

AFAICS, there is no expiry and tokens are valid until the Umami server restarts - but please note I've no clue how Next.js authentication works, so I'm happy to be corrected if anyone knows better 👍

**Comment by HoshangDEV** [comment]
For those who still looking for an answer, based on the source code, if you don't use redis (normally it's not enabled), the token has **"No Expiry"** date. 

Which means the token **remains valid until user changes password or secret key**.
